content/match-types-spec.md

```diff
@@ -0,0 +1,354 @@
+---
+layout: sip
+permalink: /sips/:title.html
+stage: implementation
+status: waiting-for-implementation
+title: SIP-56 - Proper Specification for Match Types
+---
+
+**By: Sébastien Doeraene**
+
+## History
+
+| Date          | Version            |
+|---------------|--------------------|
+| Aug 11th 2023 | Initial Draft      |
+
+## Summary
+
+Currently, match type reduction is not specified, and its implementation is by nature not specifiable.
+This is an issue because match type reduction spans across TASTy files (unlike, for example, type inference or GADTs), which can and will lead to old TASTy files not to be linked again in future versions of the compiler.
+
+This SIP proposes a proper specification for match types, which does not involve type inference.
+It is based on `baseType` computations and subtype tests involving only fully-defined types.
+That is future-proof, because `baseType` and subtyping are defined in the specification of the language.
+
+The proposed specification defines a subset of current match types that are considered legal.
+Legal match types use the new, specified reduction rules.
+Illegal match types are rejected, which is a breaking change, and can be recovered under `-source:3.3`.
+
+## Motivation
+
+Currently, match type reduction is implementation-defined.
+Matching a scrutinee type `X` against a pattern `P` with captures `ts` works as follows:
+
+1. we create new type variables for the captures `ts'` giving a pattern `P'`,
+2. we ask the compiler's `TypeComparer` (the type inference black blox) to "try and make it so" that `X <:< P'`,
+3. if it manages to do so, we get constraints for `ts'`; we then have a match, and we instantiate the body of the pattern with the received constraints for `ts`.
+
+The problem with this approach is that, by essence, type inference is an unspecified black box.
+There are *guidelines* about how it should behave in common cases, but no actual guarantee.
+This is fine everywhere else in the language, because what type inference comes up with is stored once and for all in TASTy.
+When we read TASTy files, we do not have to perform the work of type inference again; we reuse what was already computed.
+When a new version of the compiler changes type inference, it does not change what was computed and stored in TASTy files by previous versions of the compiler.
+
+For match types, this is a problem, because reduction spans across TASTy file.
+In order to guarantee compatibility, we must ensure that, for any given match type:
+
+* if it reduces in a given way in verion 1 of the compiler, it still reduces in the same way in version 2, and
+* if it decides disjointness in version 1, it still decides disjointness in version 2.
```

Comment:
terse to the point of being a bit cryptic. maybe:

```suggestion
* if it judges two types to be disjoint in version 1, it still makes the same judgment in version 2.
```

but I'm not sure what to suggest because I actually don't know what shade of meaning is intended

Reply:
I have significantly elaborated those paragraphs. Is it better now?

Reply:
That seems like a requirement in order to have the "proper specification for match types" the SIP name promises.